keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+
+Immediate allocation is useful when allocating a resource is expensive (for
+example, programming an FPGA) and the resource therefore is meant to be used by
+multiple different Pods, either in parallel or one after the other. The
+downside is that Pod resource requirements cannot be considered when choosing
+where to allocate. If a resource was allocated so that it is only available on
+one node and the Pod cannot run there because other resources like RAM or CPU
+are exhausted on that node, then the Pod cannot run elsewhere. The same applies
+to resources that are available on a certain subset of the nodes and those
+nodes are busy.
+
+Delayed allocation solves this by integrating allocation with Pod scheduling:
+an attempt to schedule a Pod triggers allocation of pending resources for nodes
+that the scheduler has deemed suitable. Scheduling the pod is then put on hold
+until all resources are allocated. This avoids scenarios where a Pod is
+permanently assigned to a node which can't fit the pod because of the pod's
+other resource requirements.
+
+When a PodTemplateSpec in an app controller spec references a ResourceClaim by
+name, all Pods created by that controller also use that name and thus share the
+resources allocated for that ResourceClaim. This might be supported by the
+resource driver and can be useful, depending on the characteristics of the
+resource.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.Resources list
+can also be a ResourceClaimTemplate. When a Pod gets created, such a template
+will be used to create a normal ResourceClaim with the Pod as owner, and then
+the normal allocation of the resource takes place.
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. For delayed allocation, a node is selected tentatively
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource is allocated, unused and was not available on a node> {
+      <randomly pick one of those resources and tell resource driver to free it>
```

Comment:
Are we willing to preempt lower-prio pods for this?

Reply:
> Are we willing to preempt lower-prio pods for this?

It's not part of the current proposal (yet). Baby steps...

Reply:
Because of the "not used" it isn't preemption. It's not going to cause a pod that is running to get killed and it a pod that has been scheduled will be able to run because its resources are not going to be taken away.

> If a resource is allocated and unused, I think it's not the responsibility of kube-scheduler to free it

The scheduler doesn't free it. It just informs the resource driver that freeing the claim would be useful to resolve a scheduling problem (pod stuck because its resources were allocated so that it cannot run on any node).

The scheduler is responsible for pod scheduling and it will detect this situation as part of its scheduling attempt for a pod, so this seems like the right place to trigger actions that will resolve the situation.